vulnerabilities/import_runner.py

```diff
@@ -24,20 +24,59 @@
 import dataclasses
 import datetime
 import logging
+from itertools import chain
 from typing import Dict
 from typing import List
 from typing import Set
 from typing import Tuple
+from typing import Optional
+from typing import Sequence
 
 import packageurl
 from django.db import DataError
+from django.core import serializers
 
 from vulnerabilities import models
 from vulnerabilities.data_source import Advisory, DataSource
 from vulnerabilities.data_source import PackageURL
 
 logger = logging.getLogger(__name__)
 
+# These _inserter classes are used to instantiate model objects.
+# Frozen dataclass  store args required to store instantiate
+# model objects, this way model objects can be hashed indirectly which
+# is required in this implementation.
+
+
+@dataclasses.dataclass(frozen=True)
+class VulnerabilityReference_inserter:
+    vulnerability: models.Vulnerability
+    reference_id: Optional[str] = ''
+    url:  Optional[str] = ''
+
+    def __post_init__(self):
+        if not any([self.reference_id, self.url]):
+            raise TypeError(
+                "VulnerabilityReference_inserter expects either reference_id or url")
+
+    def to_model_object(self):
+        return models.VulnerabilityReference(**dataclasses.asdict(self))
+
+# These _inserter classes are used to instantiate model objects.
+# Frozen dataclass  store args required to store instantiate
+# model objects, this way model objects can be hashed indirectly which
+# is required in this implementation.
+
+
+@dataclasses.dataclass(frozen=True)
+class PackageRelatedVulnerability_inserter:
+    vulnerability: models.Vulnerability
+    is_vulnerable: bool
+    package: models.Package
+
+    def to_model_object(self):
+        return models.PackageRelatedVulnerability(**dataclasses.asdict(self))
```

Comment:
Is there a need to couple these two objects?

Reply:
Yes, did you read the comments in code? There only purpose  is to `remove duplicates of model instances(these dont yet have pk ) from a list` . It's terrible approach, but I can't think of a better way to do this. There's naive way  doing it as @haikoschol  suggested above, but that will probably more or less neutralize the performance gained by bulk inserts as that approach has time complexity of O(N^2) while this dataclass approach has O(N)